### PR TITLE
Use webhook authentication from node

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/8988

This introduces a REST endpoint to back the webhook authentication type from upstream.  We use it to plumb through token scopes.  Also fixes the upstream authorization attributes so we can roundtrip through it.

I had to remove the ugorji codecs in the patch because they won't compile unless we pick down some codec changes.  We were already using the reflection path in the old code, so its not any worse.  I also can't get protobuf to generate compiling code, tagged @smarterclayton into the upstream pull.

@liggitt review?

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

File: pkg/auth/authenticator/token/remotetokenreview/removetokenreview.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package remotemaster
```
Comment (liggitt) [inline pkg/auth/authenticator/token/remotetokenreview/removetokenreview.go]: package name mismatch, filename mismatch

File: pkg/auth/authenticator/token/remotetokenreview/removetokenreview.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+func (a *Authenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+	if len(value) == 0 {
+		return nil, false, nil
+	}
+	tokenReview := &authentication.TokenReview{}
+	tokenReview.Spec.Token = value
+
+	response, err := a.authenticationClient.TokenReviews().Create(tokenReview)
+	fmt.Printf("#### 1a %#v ,%v", response, err)
```
Comment (liggitt) [inline pkg/auth/authenticator/token/remotetokenreview/removetokenreview.go]: debug

File: pkg/cmd/server/kubernetes/master_config.go [hunk 3]
```diff
@@ -197,6 +198,7 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 			PublicAddress: publicAddress,
 			ReadWritePort: port,
 
+			Authenticator:    originAuthenticator, // this is used to fulfill the tokenreviews endpoint which is used by node authentication
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/master_config.go]: does this cause the authenticator to be invoked multiple times for k8s API requests?

File: pkg/cmd/server/kubernetes/node_auth.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ import (
 	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
 )
 
-func newAuthenticator(clientCAs *x509.CertPool, anonymousConfig restclient.Config, cacheTTL time.Duration, cacheSize int) (authenticator.Request, error) {
+func newAuthenticator(authenticationClient unversionedauthentication.TokenReviewsGetter, clientCAs *x509.CertPool, cacheTTL time.Duration, cacheSize int) (authenticator.Request, error) {
 	authenticators := []oauthenticator.Request{}
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/node_auth.go]: spawn a follow up issue... we have the same issue with an invalid token and a valid client cert falling back to use the client cert

File: pkg/auth/oauth/registry/tokenauthenticator.go [hunk 5]
```diff
@@ -33,7 +33,9 @@ func (a *TokenAuthenticator) AuthenticateToken(value string) (kuser.Info, bool,
 	ctx := api.NewContext()
 
 	token, err := a.tokens.GetAccessToken(ctx, value)
+	fmt.Printf("#### 1b %v\n", token)
```
Comment (liggitt) [inline pkg/auth/oauth/registry/tokenauthenticator.go]: remove debugging?

File: pkg/cmd/server/kubernetes/master_config.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master_config.go]: > does this cause the authenticator to be invoked multiple times for k8s API requests?

Hmm... how unfortunate.  We have a similar issue with impersonation.

Comment (deads2k) [inline pkg/cmd/server/kubernetes/master_config.go]: Looks like we ignore the upstream handler chain, otherwise impersonation would fail

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
comments look minor.  Barring anything else, I'll fix up the comments, re-pick the upstream, and merge tomorrow.

**Comment by liggitt** [comment]
shape looks fine... I don't want to carry an API though, do the upstreams have LGTM?

**Comment by deads2k** [comment]
> shape looks fine... I don't want to carry an API though, do the upstreams have LGTM?

I'd be shocked if they were rejected.  You could do the user one.

**Comment by smarterclayton** [comment]
Don't merge until you have LGTM on API review upstream.

**Comment by deads2k** [comment]
Already asked Brian for it.  Of we don't get it, we'll be cloning our own and hacking it in.

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
https://github.com/openshift/origin/issues/9775  re[test]

**Comment by deads2k** [comment]
#9775 re[test]

**Comment by deads2k** [comment]
re[test] re[merge]

**Comment by deads2k** [comment]
secret mount again, re[test]

**Pull request merged**